### PR TITLE
Add torch diagonal function

### PR DESCRIPTION
**Pull request by Kaustav-Sarkar**

### CODE REVIEW

**Comment by Kaustav-Sarkar** [comment]
- [ ]  #3930

**Comment by roboFiddle** [comment]
Hi, thanks for the PR. Here's a couple changes that need to be made before I can merge this:
- Make sure that there are no linter errors. You can run `flake8` on the files you modified or check the output of the `lint/check-formatting` action on this PR
- Change the arguments to `test_frontend_function` after the `input=` line to make sure they correspond with the arguments of the `diagonal` function.

**Comment by Kaustav-Sarkar** [comment]
Sure I will make the changes !!

File: ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py [hunk 1]
```diff
@@ -187,3 +187,43 @@ def test_torch_cumsum(
...
+        as_variable_flags=as_variable,
+        with_out=True,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        fw=fw,
+        frontend="torch",
+        fn_tree="diagonal",
+        input=np.asarray(value, dtype=input_dtype),
+        offset=axis,
+        axis1=axis,
```
Comment (roboFiddle) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py:227]: - The offset, and `dim1` and `dim2` should be generated individually with hypothesis in the `@given` section. You can use `helpers.get_axis` for this.
 - These should be `dim1` and `dim2`

**Review by roboFiddle — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py (hunk 1 above)
Comment (Kaustav-Sarkar) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py:227]: Alright I will read the deep dive and try to understand in depth how the tests work. Thanks a lot !!

File: ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py [hunk 2]
```diff
@@ -187,3 +187,57 @@ def test_torch_cumsum(
...
+
+
+@given(
+    dtype_and_values=helpers.dtype_and_values(
+        available_dtypes=tuple(
+            set(ivy_np.valid_float_dtypes).intersection(
+                set(ivy_torch.valid_float_dtypes)
+            ),
+        ),
+        shape=helpers.get_shape(min_num_dims=2),
```
Comment (roboFiddle) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py:199]: this shape should also be shared with the `get_axis` calls

File: ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py [hunk 3]
```diff
@@ -187,3 +187,57 @@ def test_torch_cumsum(
...
+        as_variable_flags=as_variable,
+        with_out=True,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        fw=fw,
+        frontend="torch",
+        fn_tree="diagonal",
+        input=np.asarray(value, dtype=input_dtype),
+        offset=offset,
+        axis1=dim1,
```
Comment (roboFiddle) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_miscellaneous_ops.py:241]: the keyword parameters here should match the function signature for `diagonal`, so it should be `dim1=dim1` and `dim2=dim2`

**Review by roboFiddle — CHANGES_REQUESTED** [review]

**Comment by roboFiddle** [comment]
Looks good, I'll merge it.

**Comment by Kaustav-Sarkar** [comment]
Thanks a lot!!

**Pull request merged**